Lots of fixes and updated to the latest node version

ginnun opened the pull request:
This PR contains lots of fixes, improvements and fixes some breaking changes that comes with the latest Cardano node version (for Haskell text generation)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Structured reward-account format with on-wire encode/decode.
  - Optional Plutus V4 cost model and PlutusV4 language tagging.
  - Reusable mismatch container type and new error variants for richer diagnostics.

- **Changes**
  - Many protocol types now support ordering.
  - Transaction decoding tightened; numeric handling expanded (larger integers).
  - Error/serialization APIs now return fallible results.

- **Tests**
  - Updated to cover Plutus V4, new account format, mismatch handling, and revised decoding behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ginnun committed:
Fix Credential decoding

ginnun committed:
Match ProposalProcedure ordering with Haskell implementation

ginnun committed:
Implement HaskellDisplay for DecoderError

ginnun committed:
Use BigInt where i32 and i64 is not enough

scarmuega commented on pallas-network/src/miniprotocols/localstate/queries_v16/mod.rs:
```diff
@@ -491,16 +491,41 @@ pub enum Vote {
 
-pub type RewardAccount = Bytes;
+#[derive(Debug,  Clone, Eq, PartialEq, PartialOrd, Ord)]
+pub struct FieldedRewardAccount{
```
is the `FieldedRewardAccount` struct defined as part of some spec or is it a quality-of-life improvement for dealing with credentials?

If its the latter, I would prefer to have this in a higher layer of the stack and not as part of the codec of the miniprotocol.

ginnun replied:
@scarmuega It's there to be able to produce the same error message with Haskell codebase. Since Rust doesn't have type specialization, old-good RewardAccount (which is a Byte) can't be  used for HaskellDisplay to distinguish. 

So not sure if this is a quality-of-life improvement or not. If do you think so, I could refactor it. Just tell me what layer it should be in that case. Thanks.

ginnun committed:
fix get_current_pparams miniprotocol query

ginnun committed:
The fix for a breaking change

ginnun committed:
The fix for a breaking changes that comes with the new node

ginnun committed:
Linting

ginnun committed:
Improve Conway::Tx cbor decoding

ginnun committed:
Add new error variants and PlutusV4 support for Cardano node upgrade to 10.6.2

ginnun committed:
fix: Applied CodeRabbit's PR feedback and more

michalrus reviewed (commented):
Couple of comments, since we’re depending on this branch, and it hasn’t been been approved yet:

ginnun committed:
fix: validate reward account header per CIP-19

ginnun committed:
fix: apply PR feedback

michalrus reviewed (approved):
The CBOR tests are passing in:
- https://github.com/blockfrost/blockfrost-platform/pull/514

michalrus commented on pallas-network/src/miniprotocols/localstate/queries_v16/mod.rs:
```diff
@@ -557,16 +629,82 @@ pub enum Vote {
+    InvalidLength(usize),
+    InvalidHeaderType(u8),
+}
+
+impl fmt::Display for InvalidRewardAccount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::InvalidLength(len) => {
+                write!(f, "invalid reward account length: expected 29, got {len}")
+            }
+            Self::InvalidHeaderType(header) => {
+                write!(f, "invalid reward account header type: 0x{header:02x}")
+            }
+        }
+    }
+}
+
+impl std::error::Error for InvalidRewardAccount {}
+
+impl TryFrom<&[u8]> for FieldedRewardAccount {
+    type Error = InvalidRewardAccount;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        if bytes.len() != 29 {
+            return Err(InvalidRewardAccount::InvalidLength(bytes.len()));
+        }
+
+        // Header byte layout: [type(4 bits) | network(4 bits)]
+        // Upper nibble must be 0b1110 (key hash) or 0b1111 (script hash)
+        // per CIP-19 reward account address format.
+        let header_type = bytes[0] >> 4;
+        if !matches!(header_type, 0b1110 | 0b1111) {
+            return Err(InvalidRewardAccount::InvalidHeaderType(bytes[0]));
+        }
+
+        let network = if bytes[0] & 0b0000_0001 != 0 {
+            Network::Mainnet
+        } else {
+            Network::Testnet
+        };
```
@ginnun WDYT?

ginnun replied:
@michalrus ⚙️ cfa0c63

ginnun committed:
fix: reject non-canonical network nibble in reward account header per CIP-19

michalrus reviewed (approved):
LGTM!

scarmuega commented on pallas-primitives/src/conway/model.rs:
```diff
@@ -725,6 +731,37 @@ pub struct Tx<'b> {
 
 impl Eq for Tx<'_> {}
 
+impl<'b, C> minicbor::Decode<'b, C> for Tx<'b>
```
Could you walk me through what motivated this custom `Decode` for `Tx<'b>`?

```rust
let success = match d.decode_with(ctx) {
    Ok(v) => v,
    Err(e) if e.is_end_of_input() => true,
    Err(e) => return Err(e),
};
let auxiliary_data = match d.decode_with(ctx) {
    Ok(v) => v,
    Err(e) if e.is_end_of_input() => Nullable::Null,
    Err(e) => return Err(e),
};
```

is it required at the CDDL level for conway?

ginnun replied:
Good catch, you're right, the Conway CDDL doesn't require this.

The original idea behind the manual impl was tolerance for pre-Alonzo tx layouts ([body, witness_set, auxiliary_data / nil] and no isValid)

Decoding should fail if it's not a four-element array.

ginnun replied:
Applied in  9a83c74125738d198010dcb4d61a3595a6d4d771

ginnun committed:
fix(primitives): Conwax tx decode aligned with CDDL

ginnun committed:
Merge remote-tracking branch 'origin/main' into blockfrost

ginnun committed:
fix(hardano): wrap bare doc URLs in angle brackets

ginnun committed:
fix(validate): widen minfee_a/minfee_b to u64 across all eras

michalrus committed:
feat: support Cardano Node 11.0.1 (van Rossem / PV11) tx-submission errors

Match the Conway predicate-failure types in `cardano-ledger-conway` 1.22:
render error collections as `NonEmptySet`/`NonEmptyMap`/`NonEmpty`, fix the
`Withdrawals` wrapper, rename `StakePoolParams` fields, and `PoolMetadata`
hash / `Array` rendering.

michalrus commented:
@scarmuega, hey. 👋 I've just pushed some updates for Cardano Node 11.0.1 (van Rossem / PV11) – without them almost all of our random tests fail. After this commit, there are no failures.
